[12.0][MIG] report_py3o and report_py3o_fusion_server

lmignon opened the pull request:
This migration is based on the files of version 10.0 in order to include the latest developments made on the latest stable version of these modules.
It was not possible to duplicate some of the work done in #217 for migration to v11 because the history of this PR is not usable. In addition, #217 does not include the changes done in 10.0 to split report_py3o and report_py3o_fusion_server.
In version 10.0, the context accessible from a template was based on a `rml_parse` object. In 12.0, the context is a simple dictionary with the minimum amount of information necessary to render a template. I chose not to use the code and therefore the methods provided by `rml_parse` in 12.0 because it seems to me that they no longer made any sense in v12.0. If necessary, it will always be possible to extend the default context provided by default according to the needs that will be detected in future implementation of the module (hooks are however available to extend it to your own needs)

lmignon committed:
Move to report_py3o after import from HG

lmignon committed:
Add setup

lmignon committed:
[FIX] View inheritance may not use attribute 'string' as a selector.

lmignon committed:
[IMP] New API + Short Hearder

lmignon committed:
Remove unicode string

lmignon committed:
[IMP] Add the possiblity to get the template from an absolute path on the server

lmignon committed:
[IMP] Allow user to use local fusion

lmignon committed:
Add a simple way to extend the parser context

lmignon committed:
Remove ACL on removed model

lmignon committed:
Improve filetype selection name

lmignon committed:
Fix native flag

faide committed:
fixed missing comma in model definition

lmignon committed:
Add simple tests and fix call to fusion server

lmignon committed:
Fix skip fusion on server when the fusion is local

lmignon committed:
Improve tests

lmignon committed:
Replace openerp element by odoo

alexis-via committed:
Small usability improvements

Replace README.md by README.rst (not finished)

alexis-via committed:
Take into accounts most remarks of @lasley

Remove <data> in views
Protect import of py3o libs
Remove dep on base module
Other small changes

lmignon committed:
Replace plain SQL by orm

lmignon committed:
rename module ir_report to ir_actions_report_xml

lmignon committed:
Add more tests and fixes issues found by tests

lmignon committed:
[FIX] Check constrains only if report_type == 'py3o'

alexis-via committed:
report_py3o: full installation procedure in README

alexis-via committed:
Port report_py3o to Odoo v10

alexis-via committed:
Update version numbers

Rename __openerp__.py to __manifest__.py

alexis-via committed:
PEP8 fix

alexis-via committed:
Update README

lmignon committed:
[IMP] Escape 'False' when rendering value

lmignon committed:
[IMP] Replace old style parser by TransientModel
The goal is to improve the modularity by making the parser a true inheritable odoo model and share part of the code with the 'report' model

Conflicts:
	report_py3o/models/ir_actions_report_xml.py
	report_py3o/models/py3o_report.py
	report_py3o/tests/test_report_py3o.py

alexis-via committed:
FIX double dot in filename

alexis-via committed:
Add access to report_xml object in report

alexis-via committed:
FIX Adapt test suite to my recent commit to avoid double dot on filename extension

Commit:
[FIX] imports

lmignon committed:
[IMP] Minimizes memory consumption

Conflicts:
	report_py3o/models/py3o_report.py

Commit:
* travis.yml
* flake8

alexis-via committed:
Update and improve README instructions to setup py3o

lmignon committed:
[IMP] Allow to override/extend the way we get the fallback template

lmignon committed:
[IMP] report_py3o: Take into account print_report_name
If a Printed Report Name is set on the action report, use it as
downloaded filename
refs #133

lmignon committed:
[FIX] report_py3o: fix exception when report must be saved as attachement
The mehtod  must be called with a list of ids not with a list of browse records

oca-transbot committed:
OCA Transbot updated translations from Transifex

alexis-via committed:
[FIX] delete PDF invoice attachment on invoice back to draft (native feature now working with py3o)

lmignon committed:
[IMP] report_py3o: update authors and contributors

lmignon committed:
[IMP] report_py3o: prevent injections when retrieving the template from path

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
report_py3o: bump version following last merge

Commit:
Fix py3o freeze when using server fusion

The template file has been partially read for the parsing of the
expressions, so use the original template data instead.  When the fusion
server receives an incomplete template, we don't receive any answer
back.

Commit:
Pass option for escaping False values to server

Needs https://bitbucket.org/faide/py3o.fusion/pull-requests/1

alexis-via committed:
Add option py3o_multi_in_one for Py3o reports

oca-transbot committed:
OCA Transbot updated translations from Transifex

lmignon committed:
[FIX] report_py3o:  must return the path to the report
refs #119

oca-transbot committed:
OCA Transbot updated translations from Transifex

omar7r committed:
[IMP] report_py3o: Allows to interpret \n and \t in texts, cleans html tags too

omar7r committed:
[FIX] Deletes an unnecessary function call

omar7r committed:
[FIX] report_py3o: Removes api.one

omar7r committed:
[FIX] report_py3o: Import new dependencies in try...except

omar7r committed:
[UPD] report_py3o: Bump version

oca-transbot committed:
OCA Transbot updated translations from Transifex

hbrunn committed:
[RFR] split off the fusion server to its own module

use libreoffice for conversions in the base version. Fixes #179

jesusVMayor committed:
[FIX]report_py3o: Escape correctly html characters.

alexis-via committed:
Improve README.rst for report_py3o* modules

hbrunn committed:
[ADD] migration script to install report_py3o_fusion_server on upgrades

oca-transbot committed:
OCA Transbot updated translations from Transifex

hbrunn committed:
[FIX] fallback to libreoffice conversion when no server is configured

fixes #200

pedrobaeza committed:
[FIX] report_py3o: Bump version number

hbrunn committed:
[IMP] check for template data, not for the id

alexis-via committed:
report_py3o_fusion_server: Add support for PDF Export options of libreoffice

alexis-via committed:
report_py3o_fusion_server: Update README.rst

Fix string

alexis-via committed:
report_py3o_fusion_server: Add chronometer

Improve logs

alexis-via committed:
Fix wrong default value

SimoRubi committed:
[FIX] If this is not a py3o report, do nothing at all. In some cases the UI remains blocked otherwise

alexis-via committed:
Improve README.rst to help those that want to produce fully valid PDF/A documents

sbidoul committed:
report_py3o: mention formatLang in README

oca-travis committed:
[UPD] Update report_py3o.pot

oca-travis committed:
[UPD] Update report_py3o_fusion_server.pot

kryskool committed:
Translated using Weblate (Français)

Currently translated at 100,0% (38 of 38 strings)

Translation: reporting-engine-10.0/reporting-engine-10.0-report_py3o
Translate-URL: https://translation.odoo-community.org/projects/reporting-engine-10-0/reporting-engine-10-0-report_py3o/fr/

lmignon committed:
[MIG] report_py3o, report_py3o_fusion_server: Migration to 12.0

lmignon committed:
Tavis: Add dependencies for report_py3o and report_py3_fusion_server

lmignon committed:
Tavis: Install libreoffice requried by report_py3o

lmignon committed:
[IMP] report_py2o: cleanup js

lmignon commented:
cc @hbrunn @alexis-via @sbidoul 
Once reviewed, I'll rework the commit history to only keep one commit for the migration.

lmignon committed:
[FIX] report_py3o_fusion_server: requests.post returns bytes not str

sbidoul committed:
[ADD] setup.py

sbidoul commented:
@lmignon I pushed a fix with the setup directory

lmignon commented:
Thank you @sbidoul

Commit:
[12.0][MIG] improvement py3o_report_extender

sbejaoui committed:
Update i18n

sbejaoui commented:
@lmignon , I pushed an update of i18n. Got some issues with --i18n-overwrite.

sbidoul reviewed (approved):
Works fine (functional test).

sbidoul commented on report_py3o/README.rst:
```diff
@@ -0,0 +1,217 @@
+display_address(partner)
+    Return a formatted string of the partner's address
+formatLang(value, digits=None, date=False, date_time=False, grouping=True, monetary=False, dp=False, currency_obj=False)
+    Return a formatted numeric, monetary, date or time value according to the context language and timezone
```
@lmignon would it make sence to document `o_format_date`, `o_format_lang` here?

lmignon committed:
[FIX] report_py3o: Use the right naming convention for the render method
To determine the method to use to render a recport according to its type, the generic method 'render' defined into ir.action.report check if a method name  is defined on the model. https://github.com/odoo/odoo/blob/12.0/odoo/addons/base/models/ir_actions_report.py#L734
Therefore, we must provide this method to be compliant with what's expected by Odoo.
This change will also allows the usage of py3o template with mail_template once https://github.com/odoo/odoo/pull/30013 will be merged

lmignon committed:
[IMP] report_py3o: remove obsolete methods

njeudy reviewed (approved):
Functionnal test and works like it should :)

BT-astauder commented on report_py3o/models/py3o_report.py:
```diff
@@ -0,0 +1,360 @@
+        if len(reports_path) == 1:
+            return reports_path[0], filetype
+        if filetype == formats.FORMAT_PDF:
+            return self._merge_pdf(reports_path), formats.FORMAT_PDF
```
Great job! Thank you very much.

This function does not exists anymore, but was removed in 11.0.
Function in 10.0:
https://github.com/odoo/odoo/blob/e1767490e939233c050f796f4ae4f6db627d53c3/addons/report/models/report.py#L559
Similar functionality wor QWeb starting from 11.0:
https://github.com/odoo/odoo/blob/a7cbac305ed8645ce8ce67210d2ca5d723a891f0/odoo/addons/base/ir/ir_actions_report.py#L481

As long as this is not implemented the printing for multiple records in a listview will not work.

nicoclavier commented:
Hello and congrats on the good work, 
Any way we know when is this PR going to be merged ?
Thank you,
Nicolas

sbidoul commented on report_py3o_fusion_server/models/ir_actions_report.py:
```diff
@@ -0,0 +1,44 @@
+            if report.report_type != "py3o":
+                continue
+            is_native = Formats().get_format(report.py3o_filetype).native
+            if ((not is_native or not report.py3o_is_local_fusion) and
```
```suggestion
            if (not is_native and not report.py3o_is_local_fusion and
```

We need to change this condition, otherwise it is impossible to create new PDF reports with local fusion when `report_py3o_fusion_server` is installed. Put another way the previous condition means all new reports created must have  a fusion server defined when `report_py3o_fusion_server` which is not practical.

sbidoul replied:
@lmignon @hbrunn may I summon your opinion on this one? Thanks in advance.

hbrunn replied:
in 10 it looks like this: https://github.com/OCA/reporting-engine/blob/10.0/report_py3o_fusion_server/models/ir_actions_report_xml.py#L27 - so at least it's consistent ;-)

Can't we just leave out the is_native check entirely? If it's ie a `.doc` and there's no rendering server, the libreoffice code path will take care of it I think, right? How about some tests for the different cases?

hbrunn replied:
ah, now I understand my own code again. Check out https://github.com/OCA/reporting-engine/blob/b81219abd33d56e6068212d5bb10587b06d3db95/report_py3o_fusion_server/models/py3o_report.py#L35 - this is where those things are relevant, and the idea is that you tick the local fusion box when you have the server module installed but you want to force libreoffice conversion instead of the server. So I rectify and now this the code is fine as it is

sbidoul replied:
@hbrunn I recall now that local fusion and local format conversion is two different things. You can do local fusion, and use the py3o server to do the format conversion. Or delegate fusion to the py3o server too.

So yeah, the code is probably fine as is.

Yet we still have a problem. Let me explain our situation.
- we have an Odoo database with report_py3o_fusion_server installed
- we try to install a new module that attempts to create a new py3o PDF report; that module does not know about report_py3o_fusion_server, because it is normally used with local conversion, so it does not attach the report to a fusion server => that installation fails because the report is not native and there is no server

Not sure what to do... may be have a notion of default fusion server that is automatically set when creating new non native report when report_py3o_fusion_server is installed?

hbrunn replied:
hmmm. Make it an onchange warning maybe? Or default local_fusion to true?

lmignon replied:
@hbrunn @sbidoul The default for `py3o_is_local_fusion` is `True`. The problem is that before the split, report_py3o/report_py3O_fusion_server,  we assumed that there was no libreoffice runtime available on the machine. Therefore a py3o_server was required to transform the result of the fusion into the expected format if not native. IMO, we should have a way to detect into py3o_report that the libreoffice runtime is available and raise this error only if the runtime is not available.

lmignon replied:
@sbidoul and @hbrunn I've implemented a non blocking approach of this problem. In short:
* The constrains only fails if not local fusion and not fursion server
* We log at server startup an display a warning message on the report if the report should not work due to the unavailability of the libreoffice runtime or a fusion server
* We raise a proper error message when the libreoffice runtime is called to generate a report and is not available.

With these changes the installation of new py3o reports in non native format should work even if py3o_report_fusion_server is installed and a libreoffice runtime is not available...

lmignon committed:
[IMP] report_py3o: Document new methods available into the parser context

lmignon committed:
[FIX] report_py3o: Add missing method _merge_pdf.
This method was previously provided by Odoo and is used to merge all the reports generated if the generation is called for more than one record

alexis-via commented:
@lmignon Thanks for this great work. I'm currently working on porting https://github.com/akretion/odoo-py3o-report-templates to v12 and I have a problem: I can't use "user.something" in an expression of my template. It is important for headers and footers, where you can't access "objects" (so, in this case, I use user.company_id.partner_id.image for get the logo of the company for example)

```
 File "<string>", line 1, in <Expression "format_multiline_value(user.company_id._display_report_header([REMOVED])">
  File "/usr/local/lib/python3.6/dist-packages/genshi/template/eval.py", line 311, in lookup_name
    val = cls.undefined(name)
  File "/usr/local/lib/python3.6/dist-packages/genshi/template/eval.py", line 414, in undefined
    raise UndefinedError(key, owner=owner)
genshi.template.eval.UndefinedError: "user" not defined
```

lmignon commented:
@alexis-via I'm working on it to fix the remaining issues. 

* [X] Merge PDF  https://github.com/OCA/reporting-engine/pull/258#pullrequestreview-192189986 implemented into https://github.com/OCA/reporting-engine/pull/258/commits/213895ed0c7983d2f50eceb89675aa60f6695b57 
* [X] Constrains when py3o_report_fusion_server is installed https://github.com/OCA/reporting-engine/pull/258#pullrequestreview-197672094 implemented into https://github.com/OCA/reporting-engine/pull/258/commits/876aa27ef219a7f57fafb2ee56061570645cc35b

I add your problem 

* [X] `user`must be available into the report context. implemented into https://github.com/OCA/reporting-engine/pull/258/commits/762c52118fca4f7e7dc7ca6fa990733092dbac20

lmignon committed:
[IMP][FIX] py3o_report, py3o_report_fusion_server: Compute the availability of py3o report
Before this change it was not possible to install modules declaring py3o report into a non native format without specifying a Fusion server once the module py3o_report_fusion_server was installed. With theses changes, we now take care of the availability of the libreoffice runtime to display/log a warning message when the report is in a non native runtime.

lmignon committed:
[FIX] report_py3o: Add missing 'user' and 'lang' into the parser context

alexis-via commented:
It seems there is a pb in your last commit
```
File "/home/alexis/new_boite/dev/reporting-engine2/report_py3o/models/_py3o_parser_context.py", line 35, in __init__
    'user': self.env.user,
AttributeError: 'Py3oParserContext' object has no attribute 'env'
```

lmignon commented:
@alexis-via Sorry. should be fixed

sbidoul commented on report_py3o/models/_py3o_parser_context.py:
```diff
@@ -32,6 +32,8 @@ def __init__(self, env):
 
         self.localcontext = {
+            'user': self._env.user,
+            'lane': self._env.lang,
```
s/lane/lang/

lmignon replied:
done. Thank you

lmignon replied:
We should test all the methods provided by the parser context into the demo template to avoid regressions and also tests this part of code...

alexis-via commented:
Thanks !
On another topic, I'm having some difficulties using the new method o_format_lang() ; it's not clear to me how I should pass the first argument "env".

BTW, I would like to contribute a method for selection fields, to make it easy to display the label of a selection field in the language of the object. I'll try to find time to make a PR for that soon...

alexis-via commented:
OK, i see that you fixed o_format_lang() in your commit a few minutes ago... I can now use it. But I still have to understand how it read the lang, because it doesn't take the decimal separator that it is supposed to use...

lmignon commented:
@alexis-via The lang should be the one from the user....

alexis-via commented:
@lmignon Yes, it's what I currently have, the lang is the one of the user. Now, we want the lang to be the one of the supplier of the purchase order. Here is what I used to do:
at the beginning of the report:
```
py3o://for="obj in objects"
py3o://with="lang=obj.partner_id.lang or 'en_US'"
py3o://with="o=obj.with_context(lang=lang)"
```
then, in the report, I use "o.partner_id.name", etc..

To display the date, I use:
```
py3o://function="o_format_date(o.date_order, lang_code=lang)"
```
and it works well, I get the date in the lang of the partner of the document.

To display the total untaxed amount, I use:
```
py3o://function="o_format_lang(o.amount_untaxed, currency_obj=o.currency_id)"
```
and it displays the amount in the lang of the user... we need to fix that.

lmignon committed:
[IMP] py3o_report: Allow to specify the lang when calling o_format_lang
lang_code is already supported by o_format_date

lmignon commented:
@alexis-via Take a look at my last commit ;-) 
@GillesTephaneMeyomesse Do you call the method `o_format_lang` with named parameters? (o_format_lang(value, x=1, y=Z) ... Otherwise my last commit will break your existing reports. I would like to keep a kind of symmetry in the signature of the methods `o_format_lang` and `o_format_date`

lmignon commented:
@nicoclavier the issue with `_merge_pdf` is now fixed. Thank you for the report.

BT-kaberer commented on report_py3o/models/py3o_report.py:
```diff
@@ -302,6 +305,23 @@ def _zip_results(self, reports_path):
+        :param reports_path: list of path of pdf files
+        :returns: path of the merged pdf
+        """
+        writer = PdfFileWriter()
```
Hi. Thanks for your work!
if you did not know: there is a function in odoo core for pdf merge: https://github.com/odoo/odoo/blob/12.0/odoo/tools/pdf.py#L7
However it expects a pdf file and not a path. Feel free to include it or leave it as is. :+1:

lmignon replied:
@BT-kaberer Thank you for the reference, I'm aware of this method but I don't like it since it requires to have all the documents into memory. If you take a look at the implementation of py3o_report you'll see that we try to minimize the memory footprint of the generation process by working with files and streams ... https://github.com/OCA/reporting-engine/pull/258/files#diff-718f50f1e1aa99232a24143065702955R376

alexis-via commented:
```
File "/home/alexis/new_boite/dev/reporting-engine2/report_py3o/models/_py3o_parser_context.py", line 57, in _format_lang
    env = env.with_context(lang=lang_code)
AttributeError: 'Environment' object has no attribute 'with_context'
```

lmignon commented:
@alexis-via I'll fix it ASAP, Sorry, not tested

lmignon commented:
@alexis-via Should be fixed.
We should put a call to all the methods provided into the parser context into the demo template to be sure that these are tested by travis and avoid this kind of error....

alexis-via commented:
It works like a charm, thanks. I use:
```
py3o://function="o_format_lang(l.product_qty, dp='Product Unit of Measure', lang_code=lang)"
```

alexis-via commented:
FYI, i ported purchase_report_py3o to v12, cf https://github.com/akretion/odoo-py3o-report-templates

sbidoul commented on report_py3o/models/ir_actions_report.py:
```diff
@@ -70,6 +77,77 @@ def _get_py3o_filetypes(self):
+
+    @api.model
+    def _get_lo_bin(self):
+        lo_bin = self.env['ir.config_parameter'].get_param(
```
@lmignon a sudo is needed here. cc/ @sbejaoui

lmignon replied:
Good catch @sbidoul!
fixed

lmignon committed:
[FIX] report_py3o: Access to ir.config_parameter with sudo

sbidoul commented on report_py3o/models/py3o_report.py:
```diff
@@ -0,0 +1,384 @@
+            localcontext = self._get_parser_context(model_instance, data)
+            template.render(localcontext)
+            out_stream.seek(0)
+            tmpl_data = out_stream.read()
```
@lmignon this seek and read seems useless?

sbidoul commented:
@lmignon one more thing: the demo report is broken:
- for users with no image it produces a stack trace on `b64decode` trying to decode `False`
- for users with an image, the image does not render correctly (read error shown in the resulting file, at least with my libreoffice 6.1)

abdelhamid-attaby commented:
@lmignon Any updates in this PR?

lmignon commented:
@abdelhamid-attaby I need to find time to understand what has changed in libreoffice6 and what breaks the rendering of images. This problem seems to be a bug in the py3.template library and not related to this PR. If you do not need to insert images from Odoo data, this PR is OK.

faide commented:
> OK, i see that you fixed o_format_lang() in your commit a few minutes ago... I can now use it. But I still have to understand how it read the lang, because it doesn't take the decimal separator that it is supposed to use...

did you see the latest commits and especially this one : https://bitbucket.org/faide/py3o.template/commits/76c2209c8eaff0001a8af207c486d267959db354 that introduces babel as a way to really control your langs and formatting ?

We are still testing it because we have some differences from local tests and CI tests, but it should address your question...

And more important it deprecates the old date and number formatting techniques and will be the preferred (and soon only) way

abdelhamid-attaby commented:
@lmignon Thank you for your reply. Rendering images is very important for any templating engine. I hope you find the bugs quickly and fix it. I wish I could help but I don't have enough experience in templating engines. Sorry for that.

houzefa-abba commented on report_py3o_fusion_server/models/py3o_report.py:
```diff
@@ -0,0 +1,110 @@
+            report_xml.report_name, filetype, convert_seconds)
+        if len(model_instance) == 1:
+            self._postprocess_report(
+                result_path, model_instance.id)
```
This does not follow the signature in `report_py3o/models/py3o_report.py`:
```python
    @api.multi
    def _postprocess_report(self, model_instance, result_path):
```

Attachments don't get saved as a result. Instead the following works fine:
```suggestion
                model_instance, result_path)
```

lmignon replied:
Thank you @houzefa-abba! I've included your changes.

houzefa-abba committed:
[FIX] report_py3o: respect _postprocess_report method signature

Co-Authored-By: lmignon <laurent.mignon@acsone.eu>

lmignon commented:
>     * for users with an image, the image does not render correctly (read error shown in the resulting file, at least with my libreoffice 6.1)

The problem is into the py3o.template lib with libreoffice 6.1. I've reported the issue here https://bitbucket.org/faide/py3o.template/issues/25/images-are-no-longer-supported-with

sbidoul commented:
The image issue with libreoffice 6.1 is not blocking this PR IMO.

lmignon commented:
The image issue should be fixed by https://bitbucket.org/faide/py3o.template/pull-requests/49 (at least it work on my machine now...)
@faide Can you take a look at this fix PLZ?

lmignon commented:
The image issue seems to be a bug into libreoffice... https://bugs.documentfoundation.org/show_bug.cgi?id=123983, PR on py3o.template is a work-around.

sbidoul commented:
It looks like this PR is now good to merge. Thanks @lmignon for this huge effort.
